targets/mbedos5/tools/generate_pins.py

```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2016 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Generate pins.js for a specified target, using target definitions from the
+mbed OS source tree.
+
+It's expecting to be run from the targets/mbedos5 directory.
+"""
+
+from __future__ import print_function
+
+from pycparser import parse_file, c_ast, c_generator
+from pycparserext.ext_c_parser import GnuCParser
+
+from simpleeval import SimpleEval, DEFAULT_OPERATORS
+
+import ast
+
+import argparse
+import json
+import sys
+import os
+
+# import mbed tools
+sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'mbed-os'))
+from tools.targets import Target
+
+
+def find_file(root_dir, directories, name):
+    """
+    Find the first instance of file with name 'name' in the directory tree
+    starting with 'root_dir'.
+
+    Filter out directories that are not in directories, or do not start with
+    TARGET_.
+
+    Since this looks in (essentially )the same directories as the compiler would
+    when compiling mbed OS, we should only find one PinNames.h.
+    """
+
+    for root, dirs, files in os.walk(root_dir, topdown=True):
+        # modify dirs in place
+        dirs[:] = filter(lambda x: x in directory_labels or not x.startswith('TARGET_'), dirs)
+
+        if name in files:
+            return os.path.join(root, name)
+
+def enumerate_includes(root_dir, directories):
+    """
+    Walk through the directory tree, starting at root_dir, and enumerate all
+    valid include directories.
+    """
+    for root, dirs, files in os.walk(root_dir, topdown=True):
+        # modify dirs in place
+        dirs[:] = filter(lambda x: x in directory_labels 
```

Comment:
The directory_labels here is not passed as an argument, do we really want to use a global variable here?